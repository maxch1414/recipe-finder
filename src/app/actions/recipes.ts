"use server";

import { Recipe } from "@/lib/types";

export async function getRecipes(ingredientId: string) {
  const response = await fetch(
    `https://www.themealdb.com/api/json/v1/1/filter.php?i=${ingredientId}`
  );

  if (!response.ok) {
    throw new Error("Failed to fetch recipes");
  }

  const meals = await response.json();
  const recipes: Recipe[] = meals.meals;
  return recipes;
}
