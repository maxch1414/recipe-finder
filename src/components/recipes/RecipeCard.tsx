import React, { Dispatch, SetStateAction } from "react";
import { Card, CardHeader, CardTitle, CardContent } from "@/components/ui/card";
import { FullRecipe } from "@/lib/types";
import Image from "next/image";

type Props = {
  recipe: FullRecipe;
  setSelectedRecipe: Dispatch<SetStateAction<FullRecipe | null>>;
  openModal: () => void;
};

export const RecipeCard = ({ recipe, setSelectedRecipe, openModal }: Props) => {
  const handleClick = () => {
    setSelectedRecipe(recipe);
    openModal();
  };
  return (
    <Card
      className="overflow-hidden cursor-pointer"
      role="griditem"
      onClick={handleClick}
    >
      <CardHeader className="p-0">
        <div className="relative h-48 w-full">
          <Image
            src={recipe.strMealThumb}
            alt={recipe.strMeal}
            fill
            className="object-cover"
            sizes="(max-width: 768px) 100vw, (max-width: 1200px) 50vw, 33vw"
          />
        </div>
      </CardHeader>
      <CardContent className="p-4">
        <CardTitle className="text-xl font-semibold line-clamp-2">
          {recipe.strMeal}
        </CardTitle>
      </CardContent>
    </Card>
    // <Card
    //   className="w-full max-w-3xl overflow-hidden"
    //   role="griditem"
    //   onClick={handleClick}
    // >
    //   <CardHeader className="p-0">
    //     <div className="relative h-48 w-full">
    //       <Image
    //         src={recipe.strMealThumb}
    //         alt={recipe.strMeal}
    //         fill
    //         className="object-cover"
    //         sizes="(max-width: 768px) 100vw, (max-width: 1200px) 50vw, 33vw"
    //       />
    //     </div>
    //   </CardHeader>
    //   <CardContent className="p-4">
    //     <ScrollArea className="h-[400px] pr-4">
    //       <CardTitle className="text-2xl font-semibold mb-2">
    //         {recipe.strMeal}
    //       </CardTitle>

    //       <section className="mb-4">
    //         <h3 className="text-lg font-semibold mb-2">General Information</h3>
    //         <p>
    //           <strong>ID:</strong> {recipe.idMeal}
    //         </p>
    //         <p>
    //           <strong>Category:</strong> {recipe.strCategory}
    //         </p>
    //         <p>
    //           <strong>Area:</strong> {recipe.strArea}
    //         </p>
    //         {recipe.strDrinkAlternate && (
    //           <p>
    //             <strong>Drink Alternate:</strong> {recipe.strDrinkAlternate}
    //           </p>
    //         )}
    //         {recipe.strTags && (
    //           <div className="flex flex-wrap gap-2 mt-2">
    //             {recipe.strTags.split(",").map((tag) => (
    //               <Badge key={tag} variant="secondary">
    //                 {tag.trim()}
    //               </Badge>
    //             ))}
    //           </div>
    //         )}
    //       </section>

    //       {/* <Separator className="my-4" /> */}

    //       <section className="mb-4">
    //         <h3 className="text-lg font-semibold mb-2">Ingredients</h3>
    //         {/* <ul className="list-disc pl-5">{renderIngredients()}</ul> */}
    //       </section>

    //       {/* <Separator className="my-4" /> */}

    //       <section className="mb-4">
    //         <h3 className="text-lg font-semibold mb-2">Instructions</h3>
    //         <p>{recipe.strInstructions}</p>
    //       </section>

    //       {/* <Separator className="my-4" /> */}

    //       <section className="mb-4">
    //         <h3 className="text-lg font-semibold mb-2">
    //           Additional Information
    //         </h3>
    //         <Button variant="outline" asChild className="mb-2">
    //           <Link
    //             href={recipe.strYoutube}
    //             target="_blank"
    //             rel="noopener noreferrer"
    //           >
    //             Watch on YouTube <LinkIcon className="ml-2 h-4 w-4" />
    //           </Link>
    //         </Button>
    //       </section>
    //     </ScrollArea>
    //     <Button onClick={handleClick} className="w-full mt-4">
    //       View Full Recipe
    //     </Button>
    //   </CardContent>
    // </Card>
  );
};
