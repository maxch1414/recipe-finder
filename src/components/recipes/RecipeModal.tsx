import { Button } from "@/components/ui/button";
import {
  Dialog,
  DialogContent,
  DialogDescription,
  DialogFooter,
  DialogHeader,
  DialogTitle,
} from "@/components/ui/dialog";
import { FullRecipe } from "@/lib/types";
import { ScrollArea } from "../ui/scroll-area";
import { ExternalLink } from "lucide-react";
import { Badge } from "../ui/badge";
import Image from "next/image";

type Props = {
  recipe: FullRecipe;
  open: boolean;
  setOpen: (open: boolean) => void;
};

export function RecipeModal({ recipe, open, setOpen }: Readonly<Props>) {
  const renderIngredients = () => {
    return Array.from({ length: 20 }, (_, i) => i + 1).map((num) => {
      const ingredient = recipe[`strIngredient${num}` as keyof FullRecipe];
      const measure = recipe[`strMeasure${num}` as keyof FullRecipe];
      if (ingredient && ingredient.toString().trim() !== "") {
        return (
          <li key={num} className="mb-1">
            {ingredient}: {measure}
          </li>
        );
      }
      return null;
    });
  };

  return (
    <Dialog open={open} onOpenChange={setOpen}>
      <DialogContent className="max-w-[60vw] max-h-[90vh]">
        <DialogHeader>
          <DialogTitle className="text-2xl font-bold">
            {recipe.strMeal}
          </DialogTitle>
        </DialogHeader>
        <DialogDescription>ID: {recipe.idMeal}</DialogDescription>
        <ScrollArea className="pr-4 max-h-[calc(75vh-120px)]">
          <div className="space-y-6">
            <div className="relative h-48 w-full">
              <Image
                src={recipe.strMealThumb}
                alt={recipe.strMeal}
                fill
                className="object-cover rounded-md"
                sizes="(max-width: 768px) 100vw, (max-width: 1200px) 50vw, 33vw"
              />
            </div>

            <section>
              <p>
                <strong>Course:</strong> {recipe.strCategory}
              </p>
              <p>
                <strong>Origin:</strong> {recipe.strArea}
              </p>
              {recipe.strDrinkAlternate && (
                <p>
                  <strong>Drink Alternate:</strong> {recipe.strDrinkAlternate}
                </p>
              )}
              {recipe.strTags && (
                <div className="flex flex-wrap gap-2 mt-2">
                  <strong className="w-full">Tags:</strong>
                  {recipe.strTags.split(",").map((tag) => (
                    <Badge key={tag} variant="secondary">
                      {tag.trim()}
                    </Badge>
                  ))}
                </div>
              )}
            </section>

            <section>
              <h3 className="text-lg font-semibold mb-2">Ingredients</h3>
              <ul className="list-disc pl-5">{renderIngredients()}</ul>
            </section>

            <section>
              <h3 className="text-lg font-semibold mb-2">Instructions</h3>
              <p className="whitespace-pre-line">{recipe.strInstructions}</p>
            </section>

            <div className="space-y-2">
              <Button variant="outline" asChild className="w-full">
                <a
                  href={recipe.strYoutube}
                  target="_blank"
                  rel="noopener noreferrer"
                >
                  Watch on YouTube <ExternalLink className="ml-2 h-4 w-4" />
                </a>
              </Button>
            </div>
            <div className="space-y-2">
              <Button variant="outline" asChild className="w-full">
                <a
                  href={recipe.strSource}
                  target="_blank"
                  rel="noopener noreferrer"
                >
                  Recipe Source <ExternalLink className="ml-2 h-4 w-4" />
                </a>
              </Button>
            </div>
          </div>
        </ScrollArea>
        <DialogFooter>
          <Button
            variant="destructive"
            className="w-full"
            onClick={() => setOpen(false)}
          >
            Close
          </Button>
        </DialogFooter>
      </DialogContent>
    </Dialog>
  );
}
