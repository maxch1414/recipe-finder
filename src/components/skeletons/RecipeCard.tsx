export function RecipeCardSkeleton() {
  return (
    <div className="rounded-lg border bg-card text-card-foreground shadow-sm animate-pulse">
      <div className="relative h-48 bg-gray-200 dark:bg-gray-700 rounded-t-lg" />
      <div className="p-4 space-y-3">
        <div className="h-4 bg-gray-200 dark:bg-gray-700 rounded w-3/4" />
        <div className="h-4 bg-gray-200 dark:bg-gray-700 rounded w-1/2" />
      </div>
    </div>
  );
}
