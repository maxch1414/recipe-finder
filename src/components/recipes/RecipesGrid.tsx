import React, { useState } from "react";
import { FullRecipe } from "../../lib/types";
import { RecipeCard } from "./RecipeCard";
import { Input } from "@/components/ui/input";
import { Search } from "lucide-react";
import { RecipeModal } from "./RecipeModal";

type Props = {
  recipes: FullRecipe[];
};

export function RecipesGrid({ recipes }: Readonly<Props>) {
  const [searchQuery, setSearchQuery] = useState("");
  const [open, setOpen] = useState(false);
  const [selectedRecipe, setSelectedRecipe] = useState<FullRecipe | null>(null);

  if (!recipes.length) return;

  const filteredRecipes = recipes.filter(
    (recipe) =>
      searchQuery === "" ||
      recipe.strMeal.toLowerCase().includes(searchQuery.toLowerCase())
  );

  return (
    <div>
      {filteredRecipes.length ? (
        <div>
          <div className="relative">
            <Input
              type="text"
              placeholder="Search recipes"
              value={searchQuery}
              onChange={(e) => setSearchQuery(e.target.value)}
              className="pl-10"
            />
            <Search
              className="absolute left-3 top-1/2 transform -translate-y-1/2 text-gray-400"
              size={20}
            />
          </div>
          <div className="mt-4 grid grid-cols-1 sm:grid-cols-2 md:grid-cols-3 lg:grid-cols-4 gap-4">
            {filteredRecipes.map((recipe) => (
              <RecipeCard
                key={recipe.idMeal}
                recipe={recipe}
                setSelectedRecipe={setSelectedRecipe}
                openModal={() => setOpen(true)}
              />
            ))}
          </div>
        </div>
      ) : (
        <p className="text-center text-lg">No recipes found.</p>
      )}
      {selectedRecipe ? (
        <RecipeModal recipe={selectedRecipe} open={open} setOpen={setOpen} />
      ) : null}
    </div>
  );
}
