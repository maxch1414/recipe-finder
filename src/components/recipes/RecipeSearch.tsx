"use client";

import React, { useState } from "react";
import { FullRecipe, Ingredient } from "../../lib/types";
import { IngredientForm } from "../forms/IngredientForm";
import { RecipesGrid } from "./RecipesGrid";
import { fetchFilteredRecipes } from "@/app/actions/recipes";

type RecipeSearchProps = {
  ingredients: Ingredient[];
};

export const RecipeSearch = ({ ingredients }: RecipeSearchProps) => {
  const [recipes, setRecipes] = useState<FullRecipe[]>([]);

  const fetchRecipes = async (ingredientIds: string[]) => {
    try {
      const recipes = await fetchFilteredRecipes(ingredientIds);
      const validRecipes: FullRecipe[] = recipes.map((recipe) => ({
        ...recipe,
        ingredients: recipe.ingredients.filter(
          (ingredient): ingredient is string => ingredient !== undefined
        ),
      }));
      setRecipes(validRecipes);
    } catch (error) {
      console.error("Error fetching recipes:", error);
    }
  };

  return (
    <div>
      <IngredientForm ingredients={ingredients} onSubmit={fetchRecipes} />
      <div className="mt-4">
        <RecipesGrid recipes={recipes} />
      </div>
    </div>
  );
};
