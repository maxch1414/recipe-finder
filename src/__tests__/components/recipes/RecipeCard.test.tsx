import React from "react";
import { expect, test } from "vitest";
import { RecipeCard } from "@/components/recipes/RecipeCard";
import { render } from "@testing-library/react";
import { mockRecipes } from "@/__tests__/mocks/data/recipes";

test("Renders a recipe card correctly", () => {
  const { container } = render(<RecipeCard recipe={mockRecipes[0]} />);
  expect(container).toMatchSnapshot();
});
