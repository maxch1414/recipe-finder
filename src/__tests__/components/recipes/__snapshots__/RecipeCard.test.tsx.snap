// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Renders a recipe card correctly 1`] = `
<div>
  <div
    class="rounded-xl border bg-card text-card-foreground shadow overflow-hidden cursor-pointer"
    role="griditem"
  >
    <div
      class="flex flex-col space-y-1.5 p-0"
    >
      <div
        class="relative h-48 w-full"
      >
        <img
          alt="Test Recipe"
          class="object-cover"
          data-nimg="fill"
          decoding="async"
          loading="lazy"
          src="logo.svg"
          style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
        />
      </div>
    </div>
    <div
      class="p-4"
    >
      <h3
        class="tracking-tight text-xl font-semibold line-clamp-2"
      >
        Test Recipe
      </h3>
    </div>
  </div>
</div>
`;
